File: src/main/java/com/university/services/UserService.java
package com.university.services;

import com.university.entities.Department;
import com.university.entities.Position;
import com.university.entities.User;
import org.springframework.stereotype.Service;

import java.util.List;
import java.util.Optional;

@Service
public interface UserService {
    void saveUser(User user, boolean edit);

    void removeUser(int id);

    User getUserByEmail(String email);

    User getUserById(int id);

    List<User> getUsersByFilter(Optional<Department> department, Optional<Position> position);

    List<User> getAllUsers();

    List<User> getUsersByDepartmentAndPosition(Department department, Position position);

    List<User> getUsersByDepartment(Department department);

    List<User> getUsersByPosition(Position position);
}

File: src/main/kotlin/com/university/services/UserService.kt
package com.university.services

import com.university.entities.Department
import com.university.entities.Position
import com.university.entities.User

interface UserService {
    fun save(user: User, edit: Boolean)

    fun remove(id: Int)

    fun findByEmail(email: String): User?

    fun findById(id: Int): User

    fun findByFilter(department: Department?, position: Position?): List<User>

    fun getAll(): List<User>

    fun findByDepartmentAndPosition(department: Department, position: Position): List<User>

    fun findByDepartment(department: Department): List<User>

    fun findByPosition(position: Position): List<User>
}
